### ISSUE TITLE
GPU not detected deeplabcut-docker gui

### ISSUE THREAD

**Issue by crsleeth** [DeepLabCut/DeepLabCut#1724]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Bug description

Deeplabcut-docker gui launches but doesn't appear to be using the Nvidia GPU. Once I start analyzing videos CPU usage spikes to 100% but GPU usage is near idle. The console indicates a GPU is not detected `no NVIDIA GPU device is present: /dev/nvidia0 does not exist`. Running deeplabcut-docker bash -v /home/user:/home/user command line + `nvidia-smi` does not detect the command.

### Operating System

Ubuntu 20.04.4 LTS

### DeepLabCut version

Docker latest-gui

### DeepLabCut mode

single animal

### Device type

Tesla M40 24 GB

### Steps To Reproduce

1. Fresh install of Ubuntu 20.04.4 plus updates
2. Install Nvidia driver 510.47.03
3. `pip install deeplabcut-docker`
4. `deeplabcut-docker gui`

### Relevant log output

```shell
/usr/local/lib/python3.8/dist-packages/tensorflow/python/keras/engine/base_layer_v1.py:1692: UserWarning: `layer.apply` is deprecated and will be removed in a future version. Please use `layer.__call__` method instead.
  warnings.warn('`layer.apply` is deprecated and '
2022-03-14 21:48:57.999821: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcuda.so.1'; dlerror: libcuda.so.1: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /usr/local/nvidia/lib:/usr/local/nvidia/lib64
2022-03-14 21:48:58.000446: W tensorflow/stream_executor/cuda/cuda_driver.cc:326] failed call to cuInit: UNKNOWN ERROR (303)
2022-03-14 21:48:58.000717: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:163] no NVIDIA GPU device is present: /dev/nvidia0 does not exist
2022-03-14 21:48:58.001672: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2022-03-14 21:48:58.078094: I tensorflow/core/platform/profile_utils/cpu_utils.cc:114] CPU Frequency: 3399905000 Hz
Starting to analyze %  /app/Downloads/22099Day1.mp4
/app/Downloads  already exists!
Loading  /app/Downloads/22099Day1.mp4
Duration of video [s]:  901.5 , recorded with  30.0 fps!
Overall # of frames:  27045  found with (before cropping) frame dimensions:  1920 1080
Starting to extract posture
  0% 0/27045 [00:00<?, ?it/s]2022-03-14 21:48:59.237911: W tensorflow/core/framework/cpu_allocator_impl.cc:80] Allocation of 1061683200 exceeds 10% of free system memory.
```

### Anything else?

Running on a local display.

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/DeepLabCut/DeepLabCut/blob/master/CODE_OF_CONDUCT.md)

**Comment by MMathisLab** [comment]
@crsleeth just in you terminal, outside of docker, can you post what you get with `nvidia-smi` please?

**Comment by crsleeth** [comment]
@MMathisLab I have two workstations I'm testing with both have the same issue:

```
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 510.47.03    Driver Version: 510.47.03    CUDA Version: 11.6     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla M40 24GB      Off  | 00000000:01:00.0 Off |                    0 |
| N/A   40C    P8    17W / 250W |     65MiB / 23040MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|    0   N/A  N/A      2703      G   /usr/lib/xorg/Xorg                 57MiB |
|    0   N/A  N/A      3007      G   /usr/bin/gnome-shell                6MiB |
+-----------------------------------------------------------------------------+
```

```
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 470.103.01   Driver Version: 470.103.01   CUDA Version: 11.4     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla K40c          Off  | 00000000:03:00.0 Off |                    0 |
| 23%   36C    P8    19W / 235W |      8MiB / 11441MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   1  NVIDIA GeForce ...  Off  | 00000000:04:00.0 N/A |                  N/A |
| 50%   39C    P8    N/A /  N/A |    150MiB /   979MiB |     N/A      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|    0   N/A  N/A      1040      G   /usr/lib/xorg/Xorg                  4MiB |
+-----------------------------------------------------------------------------+
```

**Comment by greyishfox** [comment]
Hi @MMathisLab, I currently face the exact same problem. I use the same operating system and deeplabcut docker version as **crsleeth** (latest version) but with two Nvidia RTX 3090 SLI.

Are there any news regarding this issue?

Your help is very much appreciated.

![nvidia_smi](https://user-images.githubusercontent.com/91895402/162422602-1074c9a4-15db-4482-a865-e11d6b2f19ee.png)

**Comment by hejiaonju** [comment]
I am facing exactly the same issue.

Running the following code:

import tensorflow as tf
tf.config.list_physical_devices("GPU")

Returns the the following error: 

2022-07-26 22:57:55.527622: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcudart.so.11.0
2022-07-26 22:57:56.451989: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcuda.so.1'; dlerror: libcuda.so.1: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /usr/local/nvidia/lib:/usr/local/nvidia/lib64
2022-07-26 22:57:56.452008: W tensorflow/stream_executor/cuda/cuda_driver.cc:326] failed call to cuInit: UNKNOWN ERROR (303)
2022-07-26 22:57:56.452027: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:163] no NVIDIA GPU device is present: /dev/nvidia0 does not exist

**Comment by mauricev** [comment]
I can confirm [hejiaonju](https://github.com/hejiaonju)’s issue. The docker image for deeplabcut can’t utilize the GPU because it looks in the wrong place for the libcuda library and because it doesn’t see the present video hardware.

**Comment by dprotter** [comment]
I've had the same issue with the docker container (installed with the deeplabcut-docker helper). 

Not gonna lie, I tried to hack this by running the container with user privileges to create /
usr/local/nvidia/lib64 
/usr/local/nvidia/lib

and making a simlink to the libcuda.so.1 located in 
/usr/local/cuda-11.4/compat/libcuda.so.1

It didnt work....


`2022-08-30 00:11:56.982345: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcuda.so.1'; dlerror: libcuda.so.1: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /usr/local/nvidia/lib:/usr/local/nvidia/lib64
2022-08-30 00:11:56.982374: W tensorflow/stream_executor/cuda/cuda_driver.cc:326] failed call to cuInit: UNKNOWN ERROR (303)
2022-08-30 00:11:56.982396: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:163] no NVIDIA GPU device is present: /dev/nvidia0 does not exist
`

**Comment by dprotter** [comment]
i kept poking around with this.

I had some luck set a new LD_LIBRARY_PATH to include the location of libcuda.so.1 in the container. However, my GPU still wasnt found, as it did have access to /dev/nvidia0

I had luck mounting /dev as a volume. so all in all, I called the container with:

`(DLC) dprotter@dprotter-System-Product-Name:~$ export TEMP_VAR=/usr/local/nvidia/lib:/usr/local/nvidia/lib64:/usr/local/cuda-11.4/compat
(DLC) dprotter@dprotter-System-Product-Name:~$ deeplabcut-docker bash -e LD_LIBRARY_PATH=$TEMP_VAR -v /dev/:/dev/
`

and now get a classic driver mismatch error:
`>>> import tensorflow as tf
2022-08-30 19:58:55.284704: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcudart.so.11.0
tf.>>> tf.config.list_physical_devices('GPU')
2022-08-30 19:59:04.723428: I tensorflow/stream_executor/platform/default/dso_loader.cc:53] Successfully opened dynamic library libcuda.so.1
2022-08-30 19:59:04.748734: E tensorflow/stream_executor/cuda/cuda_driver.cc:328] failed call to cuInit: CUDA_ERROR_NO_DEVICE: no CUDA-capable device is detected
2022-08-30 19:59:04.748792: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:169] retrieving CUDA diagnostic information for host: 54f64c9834b9
2022-08-30 19:59:04.748824: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:176] hostname: 54f64c9834b9
2022-08-30 19:59:04.749152: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:200] libcuda reported version is: 470.57.2
2022-08-30 19:59:04.749199: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:204] kernel reported version is: 470.141.3
2022-08-30 19:59:04.749214: E tensorflow/stream_executor/cuda/cuda_diagnostics.cc:313] kernel version 470.141.3 does not match DSO version 470.57.2 -- cannot find working devices in this configuration
[]
`

tensorflow correctly finds the GPU if I'm outside of the container

**Comment by chiyu1203** [comment]
Hi everyone, I want to follow this discussion because we have a very similar problem.
We pip installed deeplabcut-docker helper package and get the helper package to pull the docker image for us.
After we built the container `deeplabcut-docker bash` , we realised that the GPU was not detected.
Therefore, we first explored what  LD_LIBRARY_PATH pointed to and set a new path for LD_LIBRARY_PATH. Then we included an extra command when we started the container  `deeplabcut-docker bash --gpus all`, somehow we could detected the GPU again. We did try to build the gui container with this extra command, however we saw this error 

> "Traceback (most recent call last):
>   File "/usr/lib/python3.8/runpy.py", line 185, in _run_module_as_main
>     mod_name, mod_spec, code = _get_module_details(mod_name, _Error)
>   File "/usr/lib/python3.8/runpy.py", line 144, in _get_module_details
>     return _get_module_details(pkg_main_name, error)
>   File "/usr/lib/python3.8/runpy.py", line 111, in _get_module_details
>     __import__(pkg_name)
>   File "/usr/local/lib/python3.8/dist-packages/deeplabcut/__init__.py", line 27, in <module>
>     from deeplabcut import generate_training_dataset
>   File "/usr/local/lib/python3.8/dist-packages/deeplabcut/generate_training_dataset/__init__.py", line 12, in <module>
>     from deeplabcut.generate_training_dataset.trainingsetmanipulation import *
>   File "/usr/local/lib/python3.8/dist-packages/deeplabcut/generate_training_dataset/trainingsetmanipulation.py", line 23, in <module>
>     from deeplabcut.pose_estimation_tensorflow import training
>   File "/usr/local/lib/python3.8/dist-packages/deeplabcut/pose_estimation_tensorflow/__init__.py", line 12, in <module>
>     from deeplabcut.pose_estimation_tensorflow.datasets import *
>   File "/usr/local/lib/python3.8/dist-packages/deeplabcut/pose_estimation_tensorflow/datasets/__init__.py", line 11, in <module>
>     from .pose_deterministic import DeterministicPoseDataset
>   File "/usr/local/lib/python3.8/dist-packages/deeplabcut/pose_estimation_tensorflow/datasets/pose_deterministic.py", line 19, in <module>
>     from deeplabcut.utils.auxfun_videos import imread, imresize
>   File "/usr/local/lib/python3.8/dist-packages/deeplabcut/utils/__init__.py", line 6, in <module>
>     from deeplabcut.utils.make_labeled_video import *
>   File "/usr/local/lib/python3.8/dist-packages/deeplabcut/utils/make_labeled_video.py", line 26, in <module>
>     import matplotlib.pyplot as plt
>   File "/usr/local/lib/python3.8/dist-packages/matplotlib/pyplot.py", line 2500, in <module>
>     switch_backend(rcParams["backend"])
>   File "/usr/local/lib/python3.8/dist-packages/matplotlib/pyplot.py", line 285, in switch_backend
>     raise ImportError(
> ImportError: Cannot load backend 'WXAgg' which requires the 'wx' interactive framework, as 'headless' is currently running
> [2022-09-07T16:06:12+0200]: Failed to launch the DLC GUI. Used args: "-e DISPLAY=unix -v /tmp/.X11-unix:/tmp/.X11-unix -v :/home/developer/.Xauthority -v /home/neuropc:/app -w /app --gpus all"

So in conclusion, we somehow managed to work around this issue by setting a new path for LD_LIBRARY_PATH and adding extra command " deeplabcut-docker bash --gpus all". However, there are extra issues with the gui mode.

**Comment by MMathisLab** [comment]
can you report the output of: 
```
cat /etc/gdm3/custom.conf | grep WaylandEnable
echo $XDG_SESSION_TYPE
echo $DISPLAY
```

**Comment by MMathisLab** [comment]
please now try running `pip install --upgrade --force-reinstall 'deeplabcut[gui]'==2.3rc1` in your env and/or docker. I will close, but please do post back!

**Comment by SteveGaukrodger** [comment]
Hi,

I'm having this problem too - I tried running `pip install --upgrade --force-reinstall 'deeplabcut[gui]'==2.3rc1` then starting again using `deeplabcut-docker gui` and running. I still see zero use of the graphics card when running `nvidia-smi`. 

Do I need to set a new LD_LIBRARY_PATH before I do this? If so, how do I set it? Just do it in the bash one and it will translate over?

**Comment by samcentanni** [comment]
Hi,

We are having the same issue as Steve. We tried pip install --upgrade --force-reinstall 'deeplabcut[gui]'==2.3rc1 and the GPU is still not being engaged (per nvidia-smi)

**Comment by MMathisLab** [comment]
On our radar to fix cc @stes , thanks @samcentanni -- should be a new container out by Jan.

**Comment by kcrisco22** [comment]
Has this been fixed?

**Comment by chiyu1203** [comment]
> Has this been fixed?

I dont know. Personally, I bumped into other issues (about compatibility issue between different version of cuda and deeplabcut docker helper package) after troubleshooting the GPU not detected issue (please check my command if you are interested in that method). Right now I have re-builded a new container with the right version of CUDA. I have also identified the same error I had before (the path issue I posted earlier). I will try the same way to troubleshoot that next week.

**Comment by chiyu1203** [comment]
> can you report the output of:
> 
> ```
> cat /etc/gdm3/custom.conf | grep WaylandEnable
> echo $XDG_SESSION_TYPE
> echo $DISPLAY
> ```

I am sorry for the delay in reply. I was not sure you were asking replying to my comment or other people's comment but when I typed those,  there is no such a file or directory  /etc/gdm3/custom.conf  in my deeplabcut docker pypi environment. And then `$XDG_SESSION_TYPE` and `echo $DISPLAY` return nothing (empty)

**Comment by thomasmichaelkane** [comment]
Has this been fixed?

I still seem to get the same issue. From the docker container when trying to use the GPU:

[Ubuntu 23.10, latest docker image, 4090]

```
2024-01-23 16:04:21.677353: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcuda.so.1'; dlerror: libcuda.so.1: cannot open shared object file: **No such file or directory; LD_LIBRARY_PATH: /usr/local/nvidia/lib:/usr/local/nvidia/lib64**
2024-01-23 16:04:21.677368: W tensorflow/stream_executor/cuda/cuda_driver.cc:326] failed call to cuInit: UNKNOWN ERROR (303)
2024-01-23 16:04:21.677379: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:163] no NVIDIA GPU device is present: /dev/nvidia0 does not exist
```

**Comment by MMathisLab** [comment]
We switched gui front ends in 2023, and no longer can support the gui from docker.

**Comment by thomasmichaelkane** [comment]
> We switched gui front ends in 2023, and no longer can support the gui from docker.

Thanks for the quick reply! I'm not trying to use the GUI though, I get this issue in docker-deeplabcut bash and notebook. Any suggestions?
